### PR TITLE
Add timeout parameter for `new_group`

### PR DESCRIPTION
**Pull request by acphile**
This PR aims to pass `timeout` parameters to `new_group` function. 

Previously, the ProcessGroups created by `new_group` does not set `timeout` parameters, which would make the communications under these ProcessGroup uses the default timeout threshold instead of what users want to set. It is crucial for failure detection.

### CODE REVIEW

File: megatron/core/parallel_state.py [hunk 1]
```diff
@@ -103,6 +103,7 @@ def initialize_model_parallel(
     context_parallel_size: int = 1,
     expert_model_parallel_size: int = 1,
     nccl_communicator_config_path: Optional[str] = None,
+    timeout = None,
```
Comment (ericharper) [inline megatron/core/parallel_state.py:106]: Can you add a type hint and docstring for this?

**Comment by ericharper** [comment]
@deepakn94 , could you help review?

**Comment by deepakn94** [comment]
Hi @acphile , thanks so much for the PR! I believe we implemented something similar in parallel: https://github.com/NVIDIA/Megatron-LM/commit/e69187bc3679ea5841030a165d587bb48b56ee77. Let me know if you think otherwise!

**Comment by acphile** [comment]
OK. Thank you for letting me know this commit. Yes, I think the intention is the same

**Pull request closed without merging**